### PR TITLE
Enable request retry using AWS SDK's default retryer

### PR DESCRIPTION
**Pull request by kyeap-vmware**
Requests to AWS can fail due to a variety of reasons beyond the control
of gpbackup. Many of these reason's can be easily resolved by retrying a
request. Reasons including, but not limited to, network instability, aws
eventual consistency, or multipart errors. Implemented using AWS SDK's
default retryer.

### CODE REVIEW

File: s3plugin/s3plugin.go [hunk 1]
```diff
@@ -171,6 +173,29 @@ func InitializeAndValidateConfig(config *PluginConfig) error {
...
+	client.DefaultRetryer
+}
+
+// ShouldRetry overrides the SDK's built in DefaultRetryer
+func (r CustomRetryer) ShouldRetry(req *request.Request) bool {
+	if r.NumMaxRetries == 0 {
+		return false
+	}
+
+	gplog.Warn("Request attempt %d failed. Next attempt in %v.\n", req.RetryCount, r.RetryRules(req))
```
Comment (jimmyyih) [inline s3plugin/s3plugin.go:187]: Is `%v` the proper specifier char?  I would think that using `%s` and maybe something like `r.RetryRules(req).String()` would be the proper format maybe?

File: s3plugin/s3plugin.go [hunk 2]
```diff
@@ -180,7 +205,7 @@ func readConfigAndStartSession(c *cli.Context) (*PluginConfig, *session.Session,
 
 	disableSSL := !ShouldEnableEncryption(config.Options.Encryption)
 
-	awsConfig := aws.NewConfig().
+	awsConfig := request.WithRetryer(aws.NewConfig(), CustomRetryer{DefaultRetryer: client.DefaultRetryer{NumMaxRetries: 5}}).
```
Comment (jimmyyih) [inline s3plugin/s3plugin.go:208]: From documentation:
```
    // MinRetryDelay is the minimum retry delay after which retry will be performed.
    // If not set, the value is 0ns.
    MinRetryDelay time.Duration

    // MinThrottleRetryDelay is the minimum retry delay when throttled.
    // If not set, the value is 0ns.
    MinThrottleDelay time.Duration

    // MaxRetryDelay is the maximum retry delay before which retry must be performed.
    // If not set, the value is 0ns.
    MaxRetryDelay time.Duration

    // MaxThrottleDelay is the maximum retry delay when throttled.
    // If not set, the value is 0ns.
    MaxThrottleDelay time.Duration
```

Shouldn't we also set some delays as well?  From my 5 seconds of looking at it, the retries seem to be instantaneous which may not work very well versus something like a simple 100ms delay.

Comment (shivzone) [reply s3plugin/s3plugin.go:208]: @jimmyyih I believe the `DefaultRetryer` (which our `CustomerRetryer` is based on) implements basic retry logic using exponential backoff

File: s3plugin/s3plugin.go [hunk 3]
```diff
@@ -171,6 +173,29 @@ func InitializeAndValidateConfig(config *PluginConfig) error {
...
+
+	gplog.Warn("Request attempt %d failed. Next attempt in %v.\n", req.RetryCount, r.RetryRules(req))
+	if req.HTTPResponse.StatusCode == 404 {
+		// 404 NoSuchKey error is possible due to AWS's eventual consistency
+		// when attempting to inspect or get a file too quickly after it was
+		// uploaded. The s3 plugin does exactly this to determine the amount of
+		// bytes uploaded. For this reason we retry 404 errors.
+		return true
+	}
+
```
Comment (shivzone) [inline s3plugin/s3plugin.go:195]: Shouldn't we also attempt to retry with this error `connection reset by peer`

**Comment by shivzone** [comment]
Not quite related to this PR, @kyeap-vmware you were talking about aws logging level that we can set which would give us additional debug logging information

File: s3plugin/s3plugin.go (hunk 1 above)
Comment (kyeap-vmware) [reply s3plugin/s3plugin.go]: I've layered 1 more commit on top that gives the AWS client a reference to gplog. This line is now removed and the AWS client will log the retry attempt.

**Comment by kyeap-vmware** [comment]
> Not quite related to this PR, @kyeap-vmware you were talking about aws logging level that we can set which would give us additional debug logging information

Yes. The next commit enables AWS client logging. you will get AWS logs when failures, retries, and the requests and replies will be recorded in the logs.

**Review by shivzone — APPROVED** [review]

**Pull request merged**